From d5e01a6b4c208bc12ed8c7b99074c5c1bece764c Mon Sep 17 00:00:00 2001
From: Sergio Prado <sergio.prado@toradex.com>
Date: Wed, 13 Jul 2022 14:52:33 -0300
Subject: [PATCH] Add support for the fdtfile variable in uEnv.txt

When TorizonCore Builder applies a device tree to a TorizonCore
image, the variable fdtfile is defined in uEnv.txt, instead of
fdt_file already supported in OSTree.

This patch adds support in ostree to this variable, to make sure its
rollback variant (fdtfile2) is created in an update, and used in a
rollback situation.

Upstream-Status: Inappropriate [TorizonCore specific]

Signed-off-by: Sergio Prado <sergio.prado@toradex.com>
---
 src/libostree/ostree-bootloader-uboot.c | 34 +++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/src/libostree/ostree-bootloader-uboot.c b/src/libostree/ostree-bootloader-uboot.c
index 7e23001e1fe7..c255fb17bf09 100644
--- a/src/libostree/ostree-bootloader-uboot.c
+++ b/src/libostree/ostree-bootloader-uboot.c
@@ -103,6 +103,36 @@ append_system_uenv (OstreeBootloaderUboot   *self,
   return TRUE;
 }
 
+static gboolean
+search_fdtfile (const char  *line,
+                void        *data,
+                GError     **error)
+{
+  g_auto(GStrv) vars = g_strsplit(line, "=", -1);
+  if (!g_strcmp0 (vars[0], "fdtfile"))
+    {
+      gchar **val = data;
+      *val = g_strdup (vars[1]);
+      return 0;
+    }
+  return 1;
+}
+
+static const char *
+uboot_config_get_fdtfile (GCancellable  *cancellable,
+                          GError        **error)
+{
+  const char *cfgfile = "/usr/lib/ostree-boot/uEnv.txt";
+  const char *val = NULL;
+
+  if (!g_file_test(cfgfile, G_FILE_TEST_EXISTS))
+    return NULL;
+
+  ot_parse_file_by_line (cfgfile, search_fdtfile, &val, cancellable, error);
+
+  return val;
+}
+
 static gboolean
 create_config_from_boot_loader_entries (OstreeBootloaderUboot     *self,
                                         int                    bootversion,
@@ -148,6 +178,10 @@ create_config_from_boot_loader_entries (OstreeBootloaderUboot     *self,
       if (val)
         g_ptr_array_add (new_lines, g_strdup_printf ("fdtdir%s=/boot%s", index_suffix, val));
 
+      val = uboot_config_get_fdtfile(cancellable, error);
+      if (val && i)
+        g_ptr_array_add (new_lines, g_strdup_printf ("fdtfile%s=%s", index_suffix, val));
+
       val = ostree_bootconfig_parser_get (config, "options");
       if (val)
         {
-- 
2.25.1

